File: backend/app/models/subscription_plan.py
from decimal import Decimal

from sqlalchemy import Boolean, Integer, JSON, Numeric, String
from sqlalchemy.orm import Mapped, mapped_column

from app.core.master_database import MasterBase


class SubscriptionPlan(MasterBase):
    __tablename__ = "subscription_plans"

    id: Mapped[int] = mapped_column(primary_key=True)
    code: Mapped[str] = mapped_column(String(40), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(80))
    monthly_price: Mapped[str | None] = mapped_column(String(30))
    annual_price: Mapped[str | None] = mapped_column(String(30))
    max_users: Mapped[int | None] = mapped_column(Integer)
    max_pdv_terminals: Mapped[int | None] = mapped_column(Integer)
    database_limit_mb: Mapped[int] = mapped_column(Integer)
    file_limit_mb: Mapped[int] = mapped_column(Integer)
    multi_company_limit: Mapped[int | None] = mapped_column(Integer)
    marketplace_listing_limit: Mapped[int | None] = mapped_column(Integer)
    api_enabled: Mapped[bool] = mapped_column(Boolean, default=False)
    priority_support: Mapped[bool] = mapped_column(Boolean, default=False)
    # Late-payment policy is owned by the plan. A disabled policy means that
    # companies on the plan never receive grace-period, fee, or interest rules
    # from the global Master finance settings.
    late_charges_enabled: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    late_fee_percent: Mapped[Decimal] = mapped_column(
        Numeric(5, 2), default=Decimal("0"), nullable=False
    )
    late_interest_daily_percent: Mapped[Decimal] = mapped_column(
        Numeric(7, 4), default=Decimal("0"), nullable=False
    )
    late_grace_days: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    default_modules: Mapped[list[str]] = mapped_column(JSON, default=list)
    active: Mapped[bool] = mapped_column(Boolean, default=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0)

File: admin_app/admin_flutter/lib/models/subscription_plan.dart
class SubscriptionPlan {
  const SubscriptionPlan({
    required this.id,
    required this.code,
    required this.name,
    this.monthlyPrice,
    this.annualPrice,
    this.maxUsers,
    required this.databaseLimitMb,
    required this.fileLimitMb,
    this.multiCompanyLimit,
    this.marketplaceListingLimit,
    required this.apiEnabled,
    required this.prioritySupport,
    required this.defaultModules,
    required this.active,
    required this.sortOrder,
  });

  final int id;
  final String code;
  final String name;
  final String? monthlyPrice;
  final String? annualPrice;
  final int? maxUsers;
  final int databaseLimitMb;
  final int fileLimitMb;
  final int? multiCompanyLimit;
  final int? marketplaceListingLimit;
  final bool apiEnabled;
  final bool prioritySupport;
  final List<String> defaultModules;
  final bool active;
  final int sortOrder;

  factory SubscriptionPlan.fromJson(Map<String, dynamic> json) {
    return SubscriptionPlan(
      id: json['id'] as int,
      code: json['code'] as String,
      name: json['name'] as String,
      monthlyPrice: json['monthly_price'] as String?,
      annualPrice: json['annual_price'] as String?,
      maxUsers: json['max_users'] as int?,
      databaseLimitMb: json['database_limit_mb'] as int,
      fileLimitMb: json['file_limit_mb'] as int,
      multiCompanyLimit: json['multi_company_limit'] as int?,
      marketplaceListingLimit: json['marketplace_listing_limit'] as int?,
      apiEnabled: json['api_enabled'] as bool? ?? false,
      prioritySupport: json['priority_support'] as bool? ?? false,
      defaultModules: (json['default_modules'] as List<dynamic>? ?? const [])
          .map((item) => item.toString())
          .toList(growable: false),
      active: json['active'] as bool? ?? true,
      sortOrder: json['sort_order'] as int? ?? 0,
    );
  }

  Map<String, dynamic> toUpdateJson() {
    return {
      'name': name,
      'monthly_price': monthlyPrice,
      'annual_price': annualPrice,
      'max_users': maxUsers,
      'database_limit_mb': databaseLimitMb,
      'file_limit_mb': fileLimitMb,
      'multi_company_limit': multiCompanyLimit,
      'marketplace_listing_limit': marketplaceListingLimit,
      'api_enabled': apiEnabled,
      'priority_support': prioritySupport,
      'default_modules': defaultModules,
      'active': active,
      'sort_order': sortOrder,
    };
  }

  Map<String, dynamic> toCreateJson() {
    return {'code': code, ...toUpdateJson()};
  }
}
